### ISSUE TITLE
BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

### ISSUE THREAD

**Issue by toriving** [toriving/papers#2]
# BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding
## 1. Paper information
BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding
Jacob Devlin, Ming-Wei Chang, Kenton Lee, Kristina Toutanova
https://arxiv.org/abs/1810.04805
Submitted on 11 Oct 2018
## 2. Abstract
BERT는 이전에 소개된 ELMo와 GPT, ULMfit 등 과 같은 pre-trained을 하여 성능을 올릴 수 있도록 하는 모델이다.  
BERT는 ELMo와 GPT와는 달리 모든 레이어에서 양방향으로 학습이 된다. (ELMo는 각 방향을 concat해서 사용하므로 얕고 독립적인 학습임.)  
간단한 구조로 많은 task에서 SOTA를 이루었다.
## 3. Introduction
pre-trained된 LM은 많은 NLP tasks에서 성능을 향상시켜왔다.   
pre-trained된 LM을 task에 적용시키는 방법은 두가지가 있는데 feature-based와 fine-tuning이다.  
ELMo와 같은 feature-based는 additional features와 같은 pre-trained된 representations을 특정 tasks에 맞는 구조에 포함시켜 사용한다.  
fine-tuning 방법은 GPT인데, 이것은 각각의 task에 맞는 parameter를 사용하고 이것을 pre-trained parameters를 이용해 fine-tuning을 한다.  
두 방법 모두 같은 목적함수로 트레이닝을 하고 모두 단방향의 LM을 사용한다.  
이러한 방법은 pre-trained된 모델의 성능을 저해하므로 우리는 fine-tuning 방법을 채택한 양방향 LM을 사용해서 성능을 올릴 것이다.  
또한 새롭게 masked language model (MLM)을 사용한다.  MLM은 인풋에서 무작위로 mask를 씌우고 이것을 예측하게 하는 방향으로 학습을 한다. 이러한 방법은 기존의 한방향으로만 학습하는 것과 달리 양쪽 방향으로 학습을 하게 해준다.   
또한 MLM과 함께 "next sentence prediction" 이라는 task를 pre-training 과정에서 함께 진행 한다. 

### Contribution 
- BiLM이 중요하다는 것을 증명할 것이다. 다른 모델과는 달리 우리는 깊은 양방향 학습을 할것이다.  GPT는 단방향이고 ELMo는 양방향이지만 독립되게 학습시킨 것을 concat하여 사용하므로 얕은 양방향이라고 할수 있다.
- 두번째는 engineering이 필요한 task-specific 한 구조 사용을 pre-train을 통해 줄인다. 또한 BERT는 문장단위와 토큰단위에서 동시에 SOTA를 얻는 최초의 모델이다.
- BERT를 통해 11개의 NLP tasks에서 SOTA를 얻었다.

## (Related Work)  
- Feature-based Approaches : ELMo - context sensitive 를 모델을 통해 추출  
- Fine-tuning Approaches : GPT - 적은 수의 parameters를 학습  
- Transfer Learning from Supervised Data : large dataset에서는 supervised learning이 더 좋다.

### 4.1 Model Architecture  
![1](https://user-images.githubusercontent.com/42434734/51098761-e78f1200-180f-11e9-8c24-ca133d477934.PNG)  

BERT는 multi-layer bidirectional Transformer encoder를 기반으로 한 모델이다. 
모델의 layer수 (Transformer blocks)을 L, hidden size를 H, self-attention heads를 A라고 칭한다.  
또한, feed-forward/filter size를 4H로 정한다.   
BERT_BASE (GPT와 같은 크기) 와 BERT_LARGE 로 구성된다.  

### 4.2 Input Representation
![2](https://user-images.githubusercontent.com/42434734/51099591-0348e700-1815-11e9-9800-5459818b9ce1.PNG)

Input representation은 single text가 들어올 수도 있고 쌍의 text가 들어 올 수 있어서 모호 할 수 있다.  
Input representation은 주어진 토큰에 대해 Token Embeddings, Segment Embeddings 그리고 Position Embeddings를 더해 만든다.

### 구체적으로는 다음과 같다.   
- 30000개의 token vocabulary를 갖는 WordPiece를 사용한다. 그리고 split word는 ##을 이용하여 표기한다.  
- 512 토큰 길이에 맞는 positional embeddings을 학습하여 사용한다.  
- 모든 sequence의 첫번째 토큰은 special classification embedding인 [CLS]를 사용한다. 해당 토큰의 마지막 hidden state는 classification task에서 사용되고 classification task가 아닐때는 무시한다.
- 문장 쌍은 하나로 묶어서 사용하지만 두가지 방법을 이용하여 문장을 구별한다. 첫번째는 special token인 [SEP]를 사용하여 분리한다. 두번째는 학습된 sentence A embedding을 첫번째 문장의 모든 토큰에 더해주고 sentence B embedding을 두번째 문장에 더해준다.
- 하나의 문장의 경우는 sentence A embedding을 사용한다.

#### 4.3.1 Task #1: Masked LM
얕은 양방향 모델인 ELMo와 단방향 모델인 GPT와 달리 깊은 양방향 모델을 사용한다.  
이를 위해 MLM을 사용하는데, 입력 토큰들을 일정 확률로 랜덤으로 마스킹하고 마스킹 된 값들을 예측하도록 한다. 그리고 마스킹한 값들만을 비교한다.  
이 방법에는 두가지 문제점이 있는데, 첫번째는 pre-training때와 fine-tuning에 격차가 생긴다는 것이다.
pre-training때는 [MASK]토큰을 사용하지만, fine-tuning때는 사용하지 않기 때문이다.  
그래서 "maked" words를 모두 [MASK]로 대체하지 않았다. 그 대신, 전체 토큰의 15%만을 masking하며, 이것은 아래와 같은 방법으로 [MASK] 로 대체된다.  
문장 : my dog is hairy --> masking 할 단어 : hairy  
- 항상 마스킹을 하는것이 아니고 아래와 같은 방법으로 한다.  
- 80%는 [MASK]로 대체한다. (my dog is hairy --> my dog is [MASK]  
- 10%는 임의의 단어로 대체한다. (my dog is hairy --> my dog is apple)  
- 10%는 그대로 둔다. (my dog is hairy --> my dog is hairy) 이렇게 함으로써 실제 관찰된것으로 편향을 줄 수 있다.  

두번째 문제는 15%만 각 배치에서 예측되는데 이 점에서는 더 많은 횟수의 트레이닝이 수렴하기까지 필요하다.   
하지만 이 점은 성능을 향상시키는데 도움이 된다.  

## 7. Related papers
None
